### PR TITLE
tests: new systemd-state tool

### PR DESCRIPTION
**Pull request by sergiocazzolato**
This is the new systemd tool that is used to create/remove systems service units on tests.

### CODE REVIEW

File: tests/lib/tools/systemd-state [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    echo "usage: create-and-start-unit <unit-name> <unit-args> [--overrides <conf>] [--persistent]"
+    echo "       stop-and-remove-unit <unit-name>"
+    echo "       wait-for-unit <unit-name>"
+    echo "       stop-units <unit-names>"
+    echo "       active-snapd-units"
+    echo ""
+    echo "Manage systemd service units and retrieve information about existing units."
+    echo ""
+    echo "COMMANDS:"
+    echo "  create-and-start-unit:   creates and starts systemd unit. In case it is set as persistent the unit survive reboots. Also it allows to overrides the configuration."
```
Comment (stolowski) [inline tests/lib/tools/systemd-state:13]:
```suggestion
    echo "  create-and-start-unit:   creates and starts systemd unit. In case it is set as persistent the unit survive reboots. Also it allows to override the configuration."
```

**Comment by stolowski** [comment]
It seem 14.04 doesn't have systemctl cat:
```
2020-10-29T14:58:03.0113921Z + systemctl cat testunit
2020-10-29T14:58:03.0114349Z Unknown operation 'cat'.
```
Perhaps just skip this particular check on 14.04?

File: tests/lib/tools/systemd-state [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    echo "  create-and-start-unit:   creates and starts systemd unit. In case it is set as persistent the unit survive reboots. Also it allows to override the configuration."
+    echo "  stop-and-remove-unit:    stops and removes a service unit"
+    echo "  wait-for-unit:           waits until a service unit is active"
+    echo "  stop-units:              stops a list of service units"
+    echo "  active-snapd-units:      retrieve a list ov active snapd units"
+}
+
+# Use like systemd_create_and_start_unit "fakestore" "$(which fakestore) -start -dir $top_dir -addr localhost:11028 $@"
+systemd_create_and_start_unit() {
+    printf '[Unit]\nDescription=Support for test %s\n[Service]\nType=simple\nExecStart=%s\n' "${SPREAD_JOB:-unknown}" "$2" > "/run/systemd/system/$1.service"
```
Comment (zyga) [inline tests/lib/tools/systemd-state:22]: May I suggest using heredoc, it'd be much easier to read and you can even discard the call to printf and use heredoc interpolation

File: tests/lib/tools/systemd-state [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    if [ -n "${3:-}" ]; then
+        echo "Environment=$3" >> "/run/systemd/system/$1.service"
```
Comment (zyga) [inline tests/lib/tools/systemd-state:24]: I would instead recommend to echo both lines from one subshell and redirect the entire output once.

File: tests/lib/tools/systemd-state [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    echo "COMMANDS:"
+    echo "  create-and-start-unit:   creates and starts systemd unit. In case it is set as persistent the unit survive reboots. Also it allows to override the configuration."
+    echo "  stop-and-remove-unit:    stops and removes a service unit"
+    echo "  wait-for-unit:           waits until a service unit is active"
+    echo "  stop-units:              stops a list of service units"
+    echo "  active-snapd-units:      retrieve a list ov active snapd units"
+}
+
+# Use like systemd_create_and_start_unit "fakestore" "$(which fakestore) -start -dir $top_dir -addr localhost:11028 $@"
+systemd_create_and_start_unit() {
```
Comment (zyga) [inline tests/lib/tools/systemd-state:21]: I kind of think this should not be a command. It's easier to just use systemd-run or creating the unit directly, without snapd-test specific wrapper that adds little features and requires everyone to look at what the implementation does anyway.

File: tests/lib/tools/systemd-state [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    printf '[Unit]\nDescription=Support for test %s\n[Service]\nType=simple\nExecStart=%s\n' "${SPREAD_JOB:-unknown}" "$2" > "/run/systemd/system/$1.service"
+    if [ -n "${3:-}" ]; then
+        echo "Environment=$3" >> "/run/systemd/system/$1.service"
+    fi
+    systemctl daemon-reload
+    systemctl start "$1"
+}
+
+# Use like systemd_create_and_start_persistent_unit "name" "start" "[Unit]\nAfter=foo"
+systemd_create_and_start_persistent_unit() {
```
Comment (zyga) [inline tests/lib/tools/systemd-state:31]: Same suggestion, I would suggest not having this command and have users mkdir / echo the unit files directly.

File: tests/lib/tools/systemd-state [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+    echo "usage: create-and-start-unit <unit-name> <unit-args> [--overrides <conf>] [--persistent]"
+    echo "       stop-and-remove-unit <unit-name>"
+    echo "       wait-for-unit <unit-name>"
+    echo "       stop-units <unit-names>"
+    echo "       active-snapd-units"
+    echo ""
+    echo "Manage systemd service units and retrieve information about existing units."
+    echo ""
+    echo "COMMANDS:"
+    echo "  create-and-start-unit:   creates and starts systemd unit. In case it is set as persistent the unit survive reboots. Also it allows to override the configuration."
```
Comment (zyga) [inline tests/lib/tools/systemd-state:13]: The description is far too long. Perhaps move this elsewhere.

File: tests/lib/tools/systemd-state [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+    systemctl start "$1"
+}
+
+# Use like systemd_create_and_start_persistent_unit "name" "start" "[Unit]\nAfter=foo"
+systemd_create_and_start_persistent_unit() {
+    printf '[Unit]\nDescription=Support for test %s\n[Service]\nType=simple\nExecStart=%s\n[Install]\nWantedBy=multi-user.target\n' "${SPREAD_JOB:-unknown}" "$2" > "/etc/systemd/system/$1.service"
+    if [ -n "${3:-}" ]; then
+        mkdir -p "/etc/systemd/system/$1.service.d"
+        # shellcheck disable=SC2059
+        printf "$3" >> "/etc/systemd/system/$1.service.d/override.conf"
```
Comment (zyga) [inline tests/lib/tools/systemd-state:36]: Why is this here and not in the main body of the unit?

File: tests/lib/tools/systemd-state (hunk 3 above)
Comment (zyga) [reply tests/lib/tools/systemd-state:24]: Also, why is Environment= set in this code and not in the code below?

**Review by zyga — COMMENTED** [review]
IMO this is too opaque and convoluted. It seems to do very little compared to what systemd offers out of the box. The extra functionality (like looping around stop) is not indicated anywhere here.

I would recommend the following:

1) remove sub-commands related to creation of systemd units. They are not really doing anything that is hard to do by hand. Doing things by hand also means that exactly what was done is visible and evident in the test. Doing it via this helper will require us to jump back and forth between the helper and test code. At some point someone will be tempted to modify the helper to do something else and then it will be a problem of leaky abstraction breaking unrelated tests.

I would keep the sub-commands that do something genuinely special, like wait-for-unit, but instead of having it duplicate repeat, I would instead write it in a way so that it is natural to use `repeat systemd-state wait-for-unit foo`. This way both tools reinforce each other and allow `repeat` options to be tweaked separately from what `systemd-state` provides. I would do something similar to `stop-units`.

Lastly there is `active-snapd-units` which feels out of place. It's very specific to snapd. Perhaps it should be in a different helper or not be a helper at all, it's just one line after all.

**Comment by sergiocazzolato** [comment]
Perhaps need to re-think this PR, closing it
